client/signup/steps/domains/index.jsx

```diff
@@ -294,7 +286,17 @@ class DomainsStep extends React.Component {
 				},
 				this.getThemeArgs()
 			),
-			Object.assign( { domainItem }, shouldHideFreePlanItem, useThemeHeadstartItem )
+
+			Object.assign(
+				{ domainItem },
+				/*
+				 * This is done to avoid enforcing a shouldHideFreePlan dependency
+				 * for steps depending on the domains component.
+				 * If shouldHideFreePlan is nundefined it will not be propegated as a dependency
+				 */
+				shouldHideFreePlan !== undefined ? { shouldHideFreePlan } : {},
```

Comment:
This is tracked under : https://github.com/Automattic/martech/issues/638

Reply:
> This will be fixed with : #45997

Thanks for taking on that refactor. I think that users going back and forth between steps could be common, so we would need this issue fixed before merging the current PR.

How do you see us proceeding on this issue? Should we wait till #45997 is reviewed and deployed? Another option  is to fix the issue without waiting for the framework level changes so that we are not blocked. The framework level change can then be reviewed separately.

If we want to fix this locally so that we are not blocked by a review for another PR, I think we can simply set the dependency like so:


```
shouldHideFreePlan  || this.getIsPlanSelectionAvailableInFlow() ? { shouldHideFreePlan } : {},
```

Hope you don't mind but I just found it easier to explain my thoughts with code.

Please let me know what you think.

Reply:
I played around with this and @niranjan-uma-shankar and the problem here is that  ``` this.getIsPlanSelectionAvailableInFlow()``` will be true for domain only flows like the following flows. 
![image](https://user-images.githubusercontent.com/3422709/94895326-d20cc080-04a8-11eb-8c25-a1e3ed161195.png)

As shown in the [implementation of getIsPlanSelectionAvailableInFlow](https://github.com/Automattic/wp-calypso/pull/45618/files#diff-df013e2a4c6a6c2b5973d50e326984f8R133-R148).  we only check if a plans aliased step appears anywhere in the flow.

So this will result in a parameter for flows where we have removed the dependency steps flows pure. 

We can try to do this change by mandating that any domains step should always provide this parameter. However I think you argued against this contradiction [already](https://github.com/Automattic/wp-calypso/pull/45618#discussion_r491830138)

However we can do something if there was an actual way to figure out, **which steps are "dead or dummy"** even if they appears in the flow schema at flows pure. This way we can truly figure out which steps are actually visible to the user. This together with checking whether a plans step occurs after the domains step we might be able to change getIsPlanSelectionAvailableInFlow](https://github.com/Automattic/wp-calypso/pull/45618/files#diff-df013e2a4c6a6c2b5973d50e326984f8R133-R148
) and provide this logic

Reply:
> and the problem here is that this.getIsPlanSelectionAvailableInFlow() will also be true for domain only flows like the following flows

Good point. Would it work if we use `forceHideFreeDomainExplainer` prop that is defined [here](https://github.com/Automattic/wp-calypso/pull/45618/files#diff-ae0bf020974ae29c034739f287afc5c5R341) to determine if the `shouldHideFreePlan` dependency should be added for the current step?

Reply:
So, it would be:

```
! forceHideFreeDomainExplainer  && ( shouldHideFreePlan  || this.getIsPlanSelectionAvailableInFlow() ) ? { shouldHideFreePlan } : {},
```

The above conditional could maybe be better organized for readability, but just quickly typed something to convey the concept.

Reply:
I did come up with this change https://github.com/Automattic/wp-calypso/pull/45618/commits/ed869f91f8dda3738b26967e57800c3bb39e8cb6. Let me know what you think @niranjan-uma-shankar

Reply:
This seems to be on the right track. I'll test it out shortly.